c error on sum type function arguments passed by reference

**V version:** 0.1.27 3664bea
**OS:** Linux 5.6.15

**What did you do?**
```v
struct CellStr { str string }

struct CellInt { itg i64 }

struct CellFloat { flt f64 }

type Cell = CellStr | CellInt | CellFloat

fn print_cell(cell &Cell) {
	match cell {
		CellStr { println(cell.str) }
		CellInt { println(cell.itg) }
		CellFloat { println(cell.flt) }
	}
}

fn main() {
    str_cell := &CellStr{ str: 'oi' }
    itg_cell := &CellInt{ itg: 1234567890 }
    flt_cell := &CellFloat{ flt: 3.141592653 }
    print_cell(str_cell)
    print_cell(itg_cell)
    print_cell(flt_cell)
}
```

**What did you expect to see?**
```
oi
1234567890
3.1415992653
 ```
**What did you see instead?**
```
/home/runner/.cache/v/main.tmp.c: In function 'print_cell':
/home/runner/.cache/v/main.tmp.c:7650:10: error: 'cell' is a pointer; did you mean to use '->'?
  if (cell.typ == 53 /* CellStr */) {
          ^
          ->
/home/runner/.cache/v/main.tmp.c:7651:31: error: 'cell' is a pointer; did you mean to use '->'?
   CellStr* it = (CellStr*)cell.obj; // ST it
                               ^
                               ->
/home/runner/.cache/v/main.tmp.c:7654:16: error: 'cell' is a pointer; did you mean to use '->'?
  }else if (cell.typ == 54 /* CellInt */) {
                ^
...
==================
(Use `v -cg` to print the entire error message)
```

--- a/vlib/v/gen/cgen.v
+++ b/vlib/v/gen/cgen.v
@@ -973,12 +973,26 @@ fn (mut g Gen) for_in(it ast.ForInStmt) {
 fn (mut g Gen) expr_with_cast(expr ast.Expr, got_type, expected_type table.Type) {
 	// cast to sum type
 	if expected_type != table.void_type {
-		if g.table.sumtype_has_variant(expected_type, got_type) {
-			got_sym := g.table.get_type_symbol(got_type)
-			got_styp := g.typ(got_type)
+		expected_is_ptr := expected_type.is_ptr()
+		expected_deref_type := if expected_is_ptr { expected_type.deref() } else { expected_type }
+		got_is_ptr := got_type.is_ptr()
+		got_deref_type := if got_is_ptr { got_type.deref() } else { got_type }
+		if g.table.sumtype_has_variant(expected_deref_type, got_deref_type) {
 			exp_styp := g.typ(expected_type)
+			got_styp := g.typ(got_type)
 			got_idx := got_type.idx()
-			if got_type.is_ptr() {
+			got_sym := g.table.get_type_symbol(got_type)
+			if expected_is_ptr && got_is_ptr {
+				exp_der_styp := g.typ(expected_deref_type)
+				g.write('/* sum type cast */ ($exp_styp) memdup(&($exp_der_styp){.obj = ')
+				g.expr(expr)
+				g.write(', .typ = $got_idx /* $got_sym.name */}, sizeof($exp_der_styp))')
+			} else if expected_is_ptr {
+				exp_der_styp := g.typ(expected_deref_type)
+				g.write('/* sum type cast */ ($exp_styp) memdup(&($exp_der_styp){.obj = memdup(&($got_styp[]) {')
+				g.expr(expr)
+				g.write('}, sizeof($got_styp)), .typ = $got_idx /* $got_sym.name */}, sizeof($exp_der_styp))')
+			} else if got_is_ptr {
 				g.write('/* sum type cast */ ($exp_styp) {.obj = ')
 				g.expr(expr)
 				g.write(', .typ = $got_idx /* $got_sym.name */}')
@@ -2222,7 +2236,9 @@ fn (mut g Gen) match_expr(node ast.MatchExpr) {
 					sym := g.table.get_type_symbol(node.cond_type)
 					// branch_sym := g.table.get_type_symbol(branch.typ)
 					if sym.kind == .sum_type {
-						g.write('.typ == ')
+						dot_or_ptr := if node.cond_type.is_ptr() { '->' } else { '.' }
+						g.write(dot_or_ptr)
+						g.write('typ == ')
 					} else if sym.kind == .interface_ {
 						// g.write('._interface_idx == _${sym.name}_${branch_sym} ')
 						g.write('._interface_idx == ')
@@ -2273,7 +2289,9 @@ fn (mut g Gen) match_expr(node ast.MatchExpr) {
 					// g.writeln('$it_type* it = ($it_type*)${tmp}.obj; // ST it')
 					g.write('\t$it_type* it = ($it_type*)')
 					g.expr(node.cond)
-					g.writeln('.obj; // ST it')
+					dot_or_ptr := if node.cond_type.is_ptr() { '->' } else { '.' }
+					g.write(dot_or_ptr)
+					g.writeln('obj; // ST it')
 					if node.var_name.len > 0 {
 						// for now we just copy it
 						g.writeln('\t$it_type* $node.var_name = it;')

--- a/vlib/v/gen/fn.v
+++ b/vlib/v/gen/fn.v
@@ -652,7 +652,9 @@ fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type table.Type) {
 		}
 		if !g.is_json_fn {
 			arg_typ_sym := g.table.get_type_symbol(arg.typ)
-			if arg_typ_sym.kind != .function {
+			expected_deref_type := if expected_type.is_ptr() { expected_type.deref() } else { expected_type }
+			is_sum_type := g.table.get_type_symbol(expected_deref_type).kind == .sum_type
+			if !((arg_typ_sym.kind == .function) || is_sum_type) {
 				g.write('(voidptr)&/*qq*/')
 			}
 		}

